### PR TITLE
Python: add HRANDFIELD command

### PR DESCRIPTION
**Pull request by aaron-congo**
*Issue #, if available:*
N/A

*Description of changes:*
- Add `HRANDFIELD` command: https://redis.io/commands/hrandfield/
- Includes value conversion: 
  - RESP2 returns a flat array of BulkString values, where even-positioned elements are keys and odd-positioned elements are values. RESP3 returns an array of array of BulkString values (in other words, a 2D-array where the inner arrays are key-value pairs). The value conversion converts the RESP2 responses to the RESP3 format.
  - Note that the returned type is a List instead of a Dict because the response may contain duplicates if the user passes in a negative `count` argument

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/python/glide/async_commands/core.py
Comment (Yury-Fridlyand) [inline python/python/glide/async_commands/core.py:723]: Should return a map I think - data should be converted in glide core - search for `ExpectedReturnType::MapOfStringToDouble` for example

File: python/python/glide/async_commands/core.py
Comment (Yury-Fridlyand) [inline python/python/glide/async_commands/core.py:742]: nit: a comment for doc

File: python/python/glide/async_commands/core.py
Comment (Yury-Fridlyand) [inline python/python/glide/async_commands/core.py:747]: make `WITHVALUES` ~great again~ a constant
It could be shared with other commands though

File: python/python/glide/async_commands/transaction.py
Comment (Yury-Fridlyand) [inline python/python/glide/async_commands/transaction.py:560]: all same comments for commands, docs and data types

File: python/python/tests/test_async_client.py
Comment (Yury-Fridlyand) [inline python/python/tests/test_async_client.py:809]: nit
```suggestion
        assert rand_field in [field, field2]
```
or even merge with previous line

File: python/python/tests/test_async_client.py
Comment (Yury-Fridlyand) [inline python/python/tests/test_async_client.py:828]: probably you can skip first `len` assert

File: python/python/tests/test_async_client.py
Comment (Yury-Fridlyand) [inline python/python/tests/test_async_client.py:831]: is there a way to compare sets?

File: python/python/tests/test_async_client.py
Comment (Yury-Fridlyand) [inline python/python/tests/test_async_client.py:837]: same

File: python/python/tests/test_async_client.py
Comment (Yury-Fridlyand) [inline python/python/tests/test_async_client.py:839]: doesnt throw? weird

File: python/python/tests/test_async_client.py
Comment (Yury-Fridlyand) [inline python/python/tests/test_async_client.py:859]: same (set comparison)

File: python/python/tests/test_async_client.py
Comment (Yury-Fridlyand) [inline python/python/tests/test_async_client.py:870]: same

File: glide-core/src/protobuf/redis_request.proto
Comment (aaron-congo) [reply glide-core/src/protobuf/redis_request.proto:139]: I thought we were supposed to take the next available number on main since it doesn't seem feasible to check all open PRs to see which numbers they are using. What's the suggested approach to determining which number to use?

Comment (Yury-Fridlyand) [reply glide-core/src/protobuf/redis_request.proto:139]: I think ~117 is the next free one. Pick any, in the worst case you will update it when resolving conflicts

File: python/python/glide/async_commands/core.py
Comment (aaron-congo) [reply python/python/glide/async_commands/core.py:723]: The problem with using a map is that a negative count value means fields can be repeated in the redis response (in other words, fields can be returned more than once in the list). If it were a map we wouldn't be able to represent repeated fields

File: python/python/glide/async_commands/core.py
Comment (aaron-congo) [reply python/python/glide/async_commands/core.py:742]: Only some of the commands in core.py have comments for the example response. I'm not sure if its necessary

File: python/python/glide/async_commands/core.py
Comment (Yury-Fridlyand) [reply python/python/glide/async_commands/core.py:723]: Oh right
But you still need data conversion to be done in glide to ensure the same format in resp2/3
```
$ redis-cli -3
127.0.0.1:6379> HRANDFIELD coin -5 WITHVALUES
1) 1) "edge"
   2) "null"
2) 1) "edge"
   2) "null"
3) 1) "edge"
   2) "null"
4) 1) "tails"
   2) "reverse"
5) 1) "edge"
   2) "null"
```
```
$ redis-cli -2
127.0.0.1:6379> HRANDFIELD coin -5 WITHVALUES
 1) "heads"
 2) "obverse"
 3) "heads"
 4) "obverse"
 5) "edge"
 6) "null"
 7) "heads"
 8) "obverse"
 9) "heads"
10) "obverse"
```

File: python/python/glide/async_commands/core.py
Comment (aaron-congo) [reply python/python/glide/async_commands/core.py:747]: zrank_withscore has a "WITHSCORE" arg and doesn't use a constant. Should I keep this as is to be consistent, change just "WITHVALUES" to be a constant, or change both to be constants?

File: python/python/glide/async_commands/transaction.py
Comment (aaron-congo) [reply python/python/glide/async_commands/transaction.py:560]: I don't understand, can you please clarify what you want changed here?

File: python/python/glide/async_commands/core.py
Comment (Yury-Fridlyand) [reply python/python/glide/async_commands/core.py:747]: Do it here or in both places - up to you

File: python/python/glide/async_commands/transaction.py
Comment (Yury-Fridlyand) [reply python/python/glide/async_commands/transaction.py:560]: It is not relevant, since we should not update the return data type

File: python/python/tests/test_async_client.py
Comment (aaron-congo) [reply python/python/tests/test_async_client.py:839]: yeah it doesn't throw. I tried executing this with redis-cli and it also returns an empty array.

File: python/python/tests/test_async_client.py
Comment (aaron-congo) [reply python/python/tests/test_async_client.py:870]: this also doesn't raise an exception and redis-cli also returns an empty array

File: python/python/tests/test_async_client.py
Comment (aaron-congo) [reply python/python/tests/test_async_client.py:856]: I can't do set comparison here because I can't convert the result to a set. The result is in the form of List[List[str]] and Python doesn't allow lists to be set keys, since they can be modified. Attempting to convert to a set gives me `TypeError: unhashable type: 'list'`

File: python/python/tests/test_async_client.py
Comment (aaron-congo) [reply python/python/tests/test_async_client.py:827]: Not a big deal either way but it might be a good idea to keep it to ensure there aren't unexpected duplicates in the results. San also has this assert in his [zrandmember PR](https://github.com/aws/glide-for-redis/pull/1238/files#diff-5db9a3bd41d647cfdba521ea248ea07d7c78bc9aaa41a26bb923d4f92a3efdd2R1091) which is similar to hrandfield

File: python/python/tests/test_async_client.py
Comment (aaron-congo) [reply python/python/tests/test_async_client.py]: yes, I updated this now

File: python/python/tests/test_async_client.py
Comment (aaron-congo) [reply python/python/tests/test_async_client.py]: I can't do set comparison here because the negative count value means the result can contain duplicates. So the resulting set could be of size one if the same field was returned every time or size two if both fields were returned at least once. But I did update to use `in` instead of `or`

File: glide-core/src/protobuf/redis_request.proto
Comment (aaron-congo) [reply glide-core/src/protobuf/redis_request.proto:139]: Sorry I'm confused, is it fine to use 95 (in other words, pick any available one and then update it when resolving conflicts) or should I be somehow checking which next one is free?

File: python/python/glide/async_commands/core.py
Comment (aaron-congo) [reply python/python/glide/async_commands/core.py:723]: Yes that is one problem I've noticed that I am not sure how to address. If resp3 returns List[List[Str]] and resp2 returns List[str], what should the signature of the function return?

Comment (aaron-congo) [reply python/python/glide/async_commands/core.py:723]: From our slack conversation sounds like we are waiting on AWS approval for what type to return

File: glide-core/src/protobuf/redis_request.proto
Comment (Yury-Fridlyand) [reply glide-core/src/protobuf/redis_request.proto:139]: 95 is already busy on upstream
You can keep it for now and update later when you'll be resolving conflicts vs upstream/main

File: python/python/glide/async_commands/core.py
Comment (Yury-Fridlyand) [inline python/python/glide/async_commands/core.py:684]:
```suggestion
            If the hash does not exist or is empty, `None` will be returned.
```

**Review by Yury-Fridlyand — APPROVED** [review]

**Comment by Yury-Fridlyand** [comment]
CI is read

**Comment by aaron-congo** [comment]
> CI is read

That is because we need to add the conversion in glide core so that the same response type is returned regardless of using RESP2 or RESP3

File: python/python/glide/async_commands/core.py
Comment (aaron-congo) [reply python/python/glide/async_commands/core.py:684]: The other docs do not surround None with angled quotes

**Comment by Yury-Fridlyand** [comment]
May be, the error is unclear

File: python/python/glide/async_commands/core.py
Comment (acarbonetto) [reply python/python/glide/async_commands/core.py:725]: We are now approved.  Please go ahead and add the data conversion.

File: CHANGELOG.md
Comment (shohamazon) [inline CHANGELOG.md:4]: add pr link
fyi, you can add it before you even add the pr

Comment (shohamazon) [reply CHANGELOG.md:4]: oh why did i think this was a pr in the upstream 👀

File: glide-core/src/client/value_conversion.rs
Comment (Yury-Fridlyand) [inline glide-core/src/client/value_conversion.rs:246]: you can move len check to line 184

File: glide-core/src/client/value_conversion.rs
Comment (Yury-Fridlyand) [inline glide-core/src/client/value_conversion.rs:245]: add a doc describing what you're doing there
and tell why a map doesn't fit

File: glide-core/src/client/value_conversion.rs
Comment (Yury-Fridlyand) [inline glide-core/src/client/value_conversion.rs:181]: when it could be empty?

File: glide-core/src/client/value_conversion.rs
Comment (Yury-Fridlyand) [inline glide-core/src/client/value_conversion.rs:279]: add a rust test for this conversion (see below in this file)

File: python/python/tests/test_async_client.py
Comment (Yury-Fridlyand) [inline python/python/tests/test_async_client.py:836]: I think you can repeat here assert from line 830 - a set drops duplicates, but you checked the lenght

File: python/python/tests/test_async_client.py
Comment (Yury-Fridlyand) [inline python/python/tests/test_async_client.py:867]: same here

File: glide-core/src/client/value_conversion.rs
Comment (aaron-congo) [reply glide-core/src/client/value_conversion.rs:246]: Wouldn't it make more sense to leave it here? If we call this function from other places in the future, we will also need to ensure the array has an even-numbered length. Having the check here instead of outside of this function means we don't need to perform this check multiple times if the function is called in multiple places.

File: glide-core/src/client/value_conversion.rs
Comment (aaron-congo) [reply glide-core/src/client/value_conversion.rs:181]: If the count value was supplied but the key doesn't exist

File: python/python/tests/test_async_client.py
Comment (aaron-congo) [reply python/python/tests/test_async_client.py:867]: I can't do that here since Python does not allow lists as hash keys

File: python/python/tests/test_async_client.py
Comment (aaron-congo) [reply python/python/tests/test_async_client.py:836]: This doesn't work because the command may return the same duplicated entry every time, or it may return different entries each time. In the first case the resulting set would be of length 1 but in the second case it would be of length 4.

File: glide-core/src/client/value_conversion.rs
Comment (Yury-Fridlyand) [inline glide-core/src/client/value_conversion.rs:248]: You can use `vec` and simplify it on line 185 too
```suggestion
fn convert_flat_array_to_key_value_pairs(array: Vec<Value>) -> RedisResult<Value> {
```

File: glide-core/src/client/value_conversion.rs
Comment (Yury-Fridlyand) [inline glide-core/src/client/value_conversion.rs:186]:
```suggestion
            Value::Array(array) if matches!(array[0], Value::BulkString(_) | Value::SimpleString(_)) => {
                convert_flat_array_to_key_value_pairs(array)
            }
```

**Review by Yury-Fridlyand — APPROVED** [review]

**Pull request merged**
